lib/server/mapper.js

```diff
@@ -126,6 +129,24 @@ class DataMapper extends EventEmitter {
               let d = {};
               d.file        = f.filepath_by_host[host]  || f.filepath_by_host["*"];
               d.accessGroup = f.access_control_group_id || '';
+
+              if (f.avh.reference) {
+                // Reference path in database is often incorrect. So use only
+                // sample name and filename from database's path, and instead
+                // substitute in actual path.
+                let ref   = f.avh.reference.split('/').slice(-5);
```

Comment:
Moving this code into function will shorten method and reduce complexity. Also we can create tests for it.